### PR TITLE
feat(react-components): Added support for storing 3D resources layers state in Url

### PR DESCRIPTION
**Pull request by pramodcog**
#### Type of change

Feat  

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/UX-1352

## Description :pencil:
3D resources layering states are stored in Url when layer state are changed

## How has this been tested? :mag:

In Storybook

## Test instructions :information_source:

- `cd react-components && yarn && yarn build && yarn storybook`
- Select `LayersButton` example
- Open the canvas of the story in a new window
- Change the 3D resources data by unselecting any loaded resources in `Layer Button` and confirm that the state changes in Url
- Can copy the Url and load it in new tab.

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [ ] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [ ] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [ ] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: react-components/src/hooks/types.ts [hunk 1]
```diff
@@ -13,3 +13,31 @@ export type Model3DEdgeProperties = {
...
+  bottom: number;
+};
+
+export type Image360LayersUrlStateParam = {
+  siteId: string;
+  applied: boolean;
+};
+
+export type CadLayersUrlStateParam = {
+  modelId: number;
```
Comment (Savokr) [inline react-components/src/hooks/types.ts:28]: `revisionId` should be used here because `modelId` can be not unique.

**Review by Savokr — APPROVED** [review]
Works really good, a few comments regarding the structure.

File: react-components/src/hooks/useUrlStateParam.tsx [hunk 2]
```diff
@@ -30,8 +35,78 @@ export const useUrlStateParam = (): UrlStateParamActions => {
...
+
+    window.history.pushState({}, '', url);
+  };
+
   return {
     getSlicerStateFromUrlParam,
-    setUrlParamOnSlicerChanged
+    setUrlParamOnSlicerChanged,
+    getLayersFromUrlParam,
+    setUrlParamOnLayersChanged
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply react-components/src/hooks/useUrlStateParam.tsx:98]: we can't use hooks like `useSearchParams` directly within React library, `useSearchParams` are meant to be used within React components that are part of a React application, and they rely on the presence of a `Router` component in the component hierarchy

File: react-components/src/components/CadModelContainer/CadModelContainer.tsx [hunk 3]
```diff
@@ -101,6 +105,18 @@ export function CadModelContainer({
...
+  function applyLayersState(model: CogniteCadModel): void {
+    if (cadLayers === undefined) {
+      return;
+    }
+    const index = viewer.models.indexOf(model);
+    cadLayers.forEach((layer) => {
+      if (layer.revisionId === revisionId && layer.index === index) {
+        model.visible = layer.applied;
+      }
+    });
```
Comment (Savokr) [inline react-components/src/components/CadModelContainer/CadModelContainer.tsx:118]: Why are you using `forEach` here, can't you just use `array.find()` to get the value of the visibility for corresponding model? Also, we were talking about adding a check for index only in the case of duplicated models but not for all models.

File: react-components/src/components/RevealToolbar/LayersButton.tsx [hunk 4]
```diff
@@ -131,9 +148,15 @@ export const LayersButton = (): ReactElement => {
       isToggled: boolean;
       isActive: boolean;
     } {
+      let visible = true;
+      image360Layers?.forEach((layer) => {
+        if (layer.siteId === image360Collection.id) {
+          visible = layer.applied;
+        }
+      });
```
Comment (Savokr) [inline react-components/src/components/RevealToolbar/LayersButton.tsx:156]: Similar to other comment related to `forEach`, can't you just use `.find` and then you don't need to define another variable and can use it in `isToggled`.

File: react-components/src/components/RevealToolbar/LayersContainer/LayersContainer.tsx [hunk 5]
```diff
@@ -7,12 +7,55 @@ import styled from 'styled-components';
...
+    });
+    setLayersUrlState({
+      cadLayers,
+      pointCloudLayers,
+      image360Layers
+    });
+  };
+
+  useEffect(() => {
+    setUrl();
```
Comment (Savokr) [inline react-components/src/components/RevealToolbar/LayersContainer/LayersContainer.tsx:57]: Function can be moved inside useEffect.

**Review by Savokr — APPROVED** [review]
Works good, a few comments!

File: react-components/src/components/RevealToolbar/LayersButton.tsx [hunk 6]
```diff
@@ -16,10 +16,17 @@ import { useReveal } from '../RevealContainer/RevealContext';
...
-export const LayersButton = (): ReactElement => {
+type LayersButtonProps = {
+  storeStateInUrl?: boolean;
+};
+
+export const LayersButton = ({ storeStateInUrl = true }: LayersButtonProps): ReactElement => {
   const viewer = useReveal();
   const revealContainerElement = useRevealContainerElement();
+  const [layersUrlState] = useLayersUrlParams();
+  const { cadLayers, pointCloudLayers, image360Layers } = layersUrlState;
```
[the remark this answers is not in the document]
Comment (pramodcog) [reply react-components/src/components/RevealToolbar/LayersButton.tsx:29]: ignoring this for now 😄

**Review by Savokr — APPROVED** [review]
Looks good!

**Pull request merged**